### PR TITLE
ne: add module

### PR DESCRIPTION
**Pull request by cwyc**
### Description

I've added a simple module to place configuration files for the ne text editor

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/rycee/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/rycee/home-manager/blob/master/doc/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/rycee/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/rycee/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [x] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

File: modules/programs/ne.nix [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      default = "";
+      example = ''
+        KEY 7f BS
+        SEQ "\x1b[1;5D" 7f
+      '';
+      description = ''
+        Keybinding file for ne
+      '';
+    };
+    defaultPreferences = mkOption {
```
Comment (rycee) [inline modules/programs/ne.nix:38]: Please separate the options with an empty line, it makes it easier to follow where they start and end. For example

```suggestion

    defaultPreferences = mkOption {
```

File: modules/programs/ne.nix [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    keybindings = mkOption {
+      type = types.lines;
+      default = "";
+      example = ''
+        KEY 7f BS
+        SEQ "\x1b[1;5D" 7f
+      '';
+      description = ''
+        Keybinding file for ne
```
Comment (rycee) [inline modules/programs/ne.nix:35]: Descriptions should have a final period to form complete sentences:
```suggestion
        Keybinding file for ne.
```

File: modules/programs/ne.nix [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+let
+  cfg = config.programs.ne;
+
+  autoPrefFiles = let
+    autoprefs = if cfg.defaultPreferences != "" then
+      cfg.automaticPreferences // {
+        ".default" = cfg.defaultPreferences;
+      } # include defaultPreferences as .default#ap
+    else
+      cfg.automaticPreferences;
```
Comment (rycee) [inline modules/programs/ne.nix:12]: A bit more idiomatic:
```suggestion
    autoprefs =
      cfg.automaticPreferences
      // optionalAttrs (cfg.defaultPreferences != "") {
        ".default" = cfg.defaultPreferences;
      }
```

File: modules/programs/ne.nix [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+      description = "Menu configuration file for ne";
+    };
+    virtualExtensions = mkOption {
+      type = types.lines;
+      default = "";
+      description = "Virtual extensions configuration file for ne";
+    };
+  };
+
+  config = mkIf cfg.enable ({
```
Comment (rycee) [inline modules/programs/ne.nix:74]: The parenthesis is unnecessary
```suggestion
  config = mkIf cfg.enable {
```

File: modules/programs/ne.nix [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+      description = ''
+        Keybinding file for ne
+      '';
+    };
+    defaultPreferences = mkOption {
+      type = types.lines;
+      default = "";
+      description = ''
+        Default preferences for ne </para><para>
+        Equivalent to programs.ne.automaticPreferences.".default"
```
Comment (rycee) [inline modules/programs/ne.nix:43]:
```suggestion
        Default preferences for ne. </para><para>
        Equivalent to <literal>programs.ne.automaticPreferences.".default"</literal>.
```

File: modules/programs/ne.nix [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    else
+      cfg.automaticPreferences;
+    keys = builtins.attrNames autoprefs;
+    gen = key: {
+      name = ".ne/${key}#ap"; # file path
+      value = {
+        text = builtins.getAttr key autoprefs;
+      }; # content, in the expected "path".text format
+    };
+  in listToAttrs (map gen keys); # assembling these into a regular record
```
Comment (rycee) [inline modules/programs/ne.nix:20]: Could use [`mapAttrs'`](https://nixos.org/nixpkgs/manual/#function-library-lib.attrsets.mapAttrs-prime). Something like
```suggestion
  in mapAttrs' gen autoprefs;
```
where `gen` is adjusted to take both the attribute name and value.

File: tests/modules/programs/ne/defprefs.nix [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  '';
+  autopref = ''
+    defined through automaticPreferences
+  '';
+in {
+  config = {
+    programs.ne = {
+      enable = true;
+      defaultPreferences = defpref;
+      automaticPreferences.".default" = autopref;
```
Comment (rycee) [inline tests/modules/programs/ne/defprefs.nix:16]: Could add some additional entry here to see that it ends up in the generated config.

**Comment by rycee** [comment]
Thanks for the contribution, I've added some comments.

**Comment by cwyc** [comment]
Thanks for the corrections, I'm still getting a hang of nix. I've force pushed the corrections. Please look at /tests/modules/programs/ne/passthroughs.nix to see if the added tests are appropriate.

**Comment by rycee** [comment]
Thanks, looks good! Rebased to master in 8f2342e13a2ddc84ed5053e2bfc393184258c686.

**Pull request closed without merging**